### ISSUE TITLE
Final Form throwing cannot read property active of undefined after submit

### ISSUE THREAD

**Issue by davidhenley** [final-form/final-form#26]
Getting error from Final Form after submission.

Stacktrace:

```
final-form.es.js:234 Uncaught (in promise) TypeError: Cannot read property 'active' of undefined
    at publishFieldState (final-form.es.js:234)
    at final-form.es.js:601
    at Array.forEach (<anonymous>)
    at notifyFieldListeners (final-form.es.js:599)
    at final-form.es.js:741
    at runValidation (final-form.es.js:587)
    at Object.initialize (final-form.es.js:740)
    at reset (final-form.es.js:822)
    at <anonymous>
publishFieldState @ final-form.es.js:234
(anonymous) @ final-form.es.js:601
notifyFieldListeners @ final-form.es.js:599
(anonymous) @ final-form.es.js:741
runValidation @ final-form.es.js:587
initialize @ final-form.es.js:740
reset @ final-form.es.js:822
Promise resolved (async)
onSubmit @ Form.js:36
apply @ _apply.js:15
baseInvoke @ _baseInvoke.js:21
apply @ _apply.js:16
(anonymous) @ _overRest.js:32
handleSubmit @ Form.js:122
callCallback @ react-dom.development.js:540
invokeGuardedCallbackDev @ react-dom.development.js:579
invokeGuardedCallback @ react-dom.development.js:436
invokeGuardedCallbackAndCatchFirstError @ react-dom.development.js:450
executeDispatch @ react-dom.development.js:834
executeDispatchesInOrder @ react-dom.development.js:856
executeDispatchesAndRelease @ react-dom.development.js:954
executeDispatchesAndReleaseTopLevel @ react-dom.development.js:965
forEachAccumulated @ react-dom.development.js:933
processEventQueue @ react-dom.development.js:1105
runEventQueueInBatch @ react-dom.development.js:3600
handleTopLevel @ react-dom.development.js:3609
handleTopLevelImpl @ react-dom.development.js:3340
batchedUpdates @ react-dom.development.js:11066
batchedUpdates @ react-dom.development.js:2323
dispatchEvent @ react-dom.development.js:3414
```

**Comment by erikras** [comment]
Can you pass in `debug = console.log` and paste here the last thing it prints out before the error?

It seems pretty odd that
```jsx
    Object.keys(fields).forEach(function (name) {
      var field = fields[name]; // <--- weird that this is null 🤔
      ...
    })
```

**Comment by davidhenley** [comment]
```
{active: undefined, dirty: false, error: undefined, errors: {…}, invalid: true, …}active: undefineddirty: falseerror: undefinederrors: customer: "Customer is required"date: "Date is required"invoice: "Invoice # required"products: undefinedtype: "Must choose pickup or replacement"__proto__: ObjectinitialValues: products: Array(1)0: {placeholder: null}length: 1__proto__: Array(0)__proto__: Objectinvalid: truepristine: truesubmitError: undefinedsubmitErrors: undefinedsubmitFailed: falsesubmitSucceeded: truesubmitting: falsevalid: falsevalidating: falsevalues: products: Array(1)0: placeholder: null__proto__: Objectlength: 1__proto__: Array(0)__proto__: Object__proto__: Object {type: {…}, customer: {…}, date: {…}, invoice: {…}, products: {…}, …}
final-form.es.js:234 Uncaught (in promise) TypeError: Cannot read property 'active' of undefined
    at publishFieldState (final-form.es.js:234)
    at final-form.es.js:601
    at Array.forEach (<anonymous>)
    at notifyFieldListeners (final-form.es.js:599)
    at final-form.es.js:741
    at runValidation (final-form.es.js:587)
    at Object.initialize (final-form.es.js:740)
    at reset (final-form.es.js:822)
    at <anonymous>
```

**Comment by davidhenley** [comment]
This is the object before hitting submit:

```
{
 "dirty": true,
 "errors": {},
 "invalid": false,
 "initialValues": {
  "products": [
   {
    "placeholder": null
   }
  ]
 },
 "pristine": false,
 "submitting": false,
 "submitFailed": false,
 "submitSucceeded": false,
 "valid": true,
 "validating": false,
 "values": {
  "products": [
   {
    "placeholder": null,
    "name": {
     "id": 159000,
     "description": "BROADBENT VV ROSE (750M)"
    },
    "cases": 1
   },
   {
    "name": {
     "id": 479405,
     "description": "CONQUISTA MALBEC (750M)"
    },
    "cases": 1
   },
   {}
  ],
  "type": "Pickup",
  "customer": {
   "id": "1545",
   "description": "FRUGAL MACDOOGAL"
  },
  "date": "2017-12-31",
  "invoice": "612045"
 }
}
```

**Comment by davidhenley** [comment]
Definitely has something to do with the FieldArray. When I take it out, everything works.

**Comment by erikras** [comment]
I'm more concerned with the _second_ thing that debug prints out, the fields.

**Comment by davidhenley** [comment]
Ahh didn't see that console.log.
Here you go:

```
{
 "type": {
  "active": false,
  "data": {},
  "lastFieldState": {
   "active": false,
   "dirty": false,
   "error": "Must choose pickup or replacement",
   "invalid": true,
   "name": "type",
   "pristine": true,
   "submitFailed": false,
   "submitSucceeded": true,
   "touched": false,
   "valid": false,
   "visited": false
  },
  "name": "type",
  "pristine": true,
  "touched": false,
  "valid": true,
  "validators": {},
  "visited": false,
  "error": "Must choose pickup or replacement"
 },
 "customer": {
  "active": false,
  "data": {},
  "lastFieldState": {
   "active": false,
   "dirty": false,
   "error": "Customer is required",
   "invalid": true,
   "name": "customer",
   "pristine": true,
   "submitFailed": false,
   "submitSucceeded": true,
   "touched": false,
   "valid": false,
   "visited": false
  },
  "name": "customer",
  "pristine": true,
  "touched": false,
  "valid": true,
  "validators": {},
  "visited": false,
  "error": "Customer is required"
 },
 "date": {
  "active": false,
  "data": {},
  "lastFieldState": {
   "active": false,
   "dirty": false,
   "error": "Date is required",
   "invalid": true,
   "name": "date",
   "pristine": true,
   "submitFailed": false,
   "submitSucceeded": true,
   "touched": false,
   "valid": false,
   "visited": false
  },
  "name": "date",
  "pristine": true,
  "touched": false,
  "valid": true,
  "validators": {},
  "visited": false,
  "error": "Date is required"
 },
 "invoice": {
  "active": false,
  "data": {},
  "lastFieldState": {
   "active": false,
   "dirty": false,
   "error": "Invoice # required",
   "invalid": true,
   "name": "invoice",
   "pristine": true,
   "submitFailed": false,
   "submitSucceeded": true,
   "touched": false,
   "valid": false,
   "visited": false
  },
  "name": "invoice",
  "pristine": true,
  "touched": false,
  "valid": true,
  "validators": {},
  "visited": false,
  "error": "Invoice # required"
 },
 "products": {
  "active": false,
  "data": {},
  "initial": [
   {
    "placeholder": null
   }
  ],
  "lastFieldState": {
   "active": false,
   "dirty": false,
   "initial": [
    {
     "placeholder": null
    }
   ],
   "invalid": false,
   "length": 1,
   "name": "products",
   "pristine": true,
   "submitFailed": false,
   "submitSucceeded": true,
   "touched": false,
   "valid": true,
   "value": [
    {
     "placeholder": null
    }
   ],
   "visited": false
  },
  "name": "products",
  "pristine": true,
  "touched": false,
  "valid": true,
  "value": [
   {
    "placeholder": null
   }
  ],
  "validators": {},
  "visited": false
 },
 "products[0].name": {
  "active": false,
  "data": {},
  "lastFieldState": {
   "active": false,
   "dirty": false,
   "invalid": false,
   "name": "products[0].name",
   "pristine": true,
   "submitFailed": false,
   "submitSucceeded": true,
   "touched": false,
   "valid": true,
   "visited": false
  },
  "name": "products[0].name",
  "pristine": true,
  "touched": false,
  "valid": true,
  "validators": {},
  "visited": false
 },
 "products[0].cases": {
  "active": false,
  "data": {},
  "lastFieldState": {
   "active": false,
   "dirty": false,
   "invalid": false,
   "name": "products[0].cases",
   "pristine": true,
   "submitFailed": false,
   "submitSucceeded": true,
   "touched": false,
   "valid": true,
   "visited": false
  },
  "name": "products[0].cases",
  "pristine": true,
  "touched": false,
  "valid": true,
  "validators": {},
  "visited": false
 },
 "products[0].bottles": {
  "active": false,
  "data": {},
  "lastFieldState": {
   "active": false,
   "dirty": false,
   "invalid": false,
   "name": "products[0].bottles",
   "pristine": true,
   "submitFailed": false,
   "submitSucceeded": true,
   "touched": false,
   "valid": true,
   "visited": false
  },
  "name": "products[0].bottles",
  "pristine": true,
  "touched": false,
  "valid": true,
  "validators": {},
  "visited": false
 },
 "products[0].reason": {
  "active": false,
  "data": {},
  "lastFieldState": {
   "active": false,
   "dirty": false,
   "invalid": false,
   "name": "products[0].reason",
   "pristine": true,
   "submitFailed": false,
   "submitSucceeded": true,
   "touched": false,
   "valid": true,
   "visited": false
  },
  "name": "products[0].reason",
  "pristine": true,
  "touched": false,
  "valid": true,
  "validators": {},
  "visited": false
 },
 "products[0].info": {
  "active": false,
  "data": {},
  "lastFieldState": {
   "active": false,
   "dirty": false,
   "invalid": false,
   "name": "products[0].info",
   "pristine": true,
   "submitFailed": false,
   "submitSucceeded": true,
   "touched": false,
   "valid": true,
   "visited": false
  },
  "name": "products[0].info",
  "pristine": true,
  "touched": false,
  "valid": true,
  "validators": {},
  "visited": false
 }
}
```

**Comment by davidhenley** [comment]
Does it have something to do with putting a null placeholder?

**Comment by erikras** [comment]
> Does it have something to do with putting a null placeholder?

No. I think I have enough here to reproduce it. I'll get back to you in a bit...

**Comment by erikras** [comment]
Reproduced!

[![Edit 🏁 React Final Form - Field Array Reset Bug](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/yk34xow83j)

**Comment by erikras** [comment]
Published fix in [`react-final-form-arrays v1.0.1`](https://github.com/final-form/react-final-form-arrays/releases/tag/v1.0.1).

**Comment by Fer0x** [comment]
Got this error without using `react-final-form-arrays` by calling `form.reset({ empty: isEmpty })` in `onSubmit`

**Comment by erikras** [comment]
@Fer0x Sandbox?

**Comment by thulani-chivandikwa** [comment]
I am also getting this error on form.reset and on updating the changed field array both with array mutators and a call for formApi.change

**Comment by sedx** [comment]
Hi! Get same error ^
@erikras There is sandbox https://codesandbox.io/s/funny-darkness-ie3qm?fontsize=14
For reproduce click `add` then `save`

**Comment by erikras** [comment]
Thanks, @sedx. Changing `reset()` to `setTImeout(reset)` works, but this is clearly a bug.

**Comment by erikras** [comment]
@sedx What's happening here is that the `reset()` is causing the form to rerender without the fields, and _then_ the `Promise` resolves, causing a notification to go out to those fields, but they are no longer registered.

I think the problem is in how you're simulating your backend call on this. Because `Promise.resolve()` is _synchronous_, but Final Form is assuming that, if `onSubmit` is returning a `Promise`, it must be async... If you put the `reset()` call inside the `Promise` callback, it works fine.

[![Edit priceless-lake-youom](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/priceless-lake-youom?fontsize=14)

All that said, perhaps Final Form should do a `field === undefined` check and abort the notification. That seems like a reasonable thing for it to do.

**Comment by sedx** [comment]
@erikras Hmmm. Yep, sandbox has this issue (especially that I not return Promise 😅), but in my code (when I call actual server) I return promise and have same issue.

I update sandbox. Add call to external service with 1s latency, return promise from `onSubmit` handler (disable save button during submitting to check that all is correct) and issue still present.

Actually, in your sandbox, you resolve promise after form reset, so it's equivalent that I reset from before  get response from the remote service. Change this construction to something like this:
```
return new Promise(resolve => {
      // simulate 2s server response delay
      setTimeout(resolve, 2e3);
    })
       // when response received
      .then(() => {
        reset();
        alert(successMessage);
      })
      .catch(e => {
        console.log(e);
        alert(errorMessage);
      });
```
and you get the error.

**Comment by erikras** [comment]
Excellent reporting. 👍 I've found the bug. It's one of these fun obscure ones. Release coming shortly...

**Comment by erikras** [comment]
Published fix in [`v4.13.1`](https://github.com/final-form/final-form/releases/tag/v4.13.1).

Confirmed that your sandbox works now with `v4.13.1`. 👍
